### DISCUSSION TITLE
[technical-sharing] how do you add click event to clickhouse?

### DISCUSSION THREAD

**Discussion by hiepxanh** [plausible/analytics#1159] [Anything Else]
Thank you for your wonderful application, It very good, I'm currently using Plausible for my application for analytic and it smooth, just trying to integrate daily view count for my book and book chapter, I'm develop a custom system, and I wonder how you transfer count data to clickhouse? Im so noob, and dont know where to ask. it really new to me and the internet is wide, I already read the clickhouse.tech document, it good but not enough. I have few question, if you dont like it, I can close this discussion immediately:
1) clickhouse limit 1 insert/ 1 second, how you insert data to clickhouse? I can't insert data directly to clickhouse, I already read your source but not see any kind of rabitmq, should I insert data to postgresql first then using aggregation_job as picture to insert calculated data to clickhouse? what kind of system do you insert?
2) should I using raw data insert to clickhouse? should I delete it after time? or using TTL or somekind of auto delete partition data? I see some kind of mergeTree, and you are using CollapsingMergeTree, how do you do that by what? ( can you send me that code line? the github repo is so big)

thank you for sharing your knowlege, hope you dont mind to write a few line, or you can give me some blog or book I can read. it will help me alot

![image](https://user-images.githubusercontent.com/16970990/123539810-b7305980-d765-11eb-8d38-7e7d4cb2c12e.png)

**Comment by hiepxanh** [comment]
@ukutaht

**Comment by ukutaht** [comment]
Hey! Happy to share

1. You can insert the data in large batches. The 1 insert per second suggestion applies to insert _statements_ but every insert statement can have a large number of events. You can try inserting 10k events per second in a single insert statement and that should work fine.
2. Yes you can just insert raw data, no need to add a TTL.

**Reply by hiepxanh** [reply]
Thank you so much, hi I just read all the blog and slideshare from  https://altinity.com/blog which is very good resouce.  there is really little tutorial or practice about clickhouse. 
1) Do you have any suggest about where should I learn more about it? (I already read the clickhouse.tech and learn the only one course on udemy)
2) I see from this slideshare https://www.slideshare.net/Altinity/analytics-at-speed-introduction-to-clickhouse-and-common-use-cases-by-mikhail-filimonov-altinity (on the slide 27)  Since the limitation is 1 query/second, It suggest using one of this technology to push data. I only see the choice is:
- or using postgresql as data lake then push to Clickhouse
- or using Kafka as data pipeline and stream it to Clickhouse
but I already read your postgresql table and research in your github repo, I dont see any signanture of Kafka, I'm really have a gap between javascript client website and clickhouse, how do you push it from website to clickhouse? which kind of tool are you using?

![image](https://user-images.githubusercontent.com/16970990/123651269-86c7e880-d855-11eb-84ff-ba6ef8c89085.png)

![image](https://user-images.githubusercontent.com/16970990/123651019-554f1d00-d855-11eb-9038-76f59c7a73be.png)

**Reply by hiepxanh** [reply]
it could be alot of query each page views on the rush hour, how can you push alot of that  view count, event.v..v... to the clickhouse? I mean I know I can insert with large batch, but with which tool are you using? Postgresql? Kafka? 

hope you can share more about it 😄

**Reply by ukutaht** [reply]
Our Elixir/Phoenix web server buffers the events up and sends them in large batches. The code is here: https://github.com/plausible/analytics/blob/master/lib/plausible/event/write_buffer.ex

**Reply by hiepxanh** [reply]
Woa that so sweet, I really appreciate that, thank you sir
